pkg/ha/doc.go

```diff
@@ -0,0 +1,146 @@
+//   - the DM-worker READ schedule operations and obey them.
+// In other words, behaviors of the cluster are clear, that are decisions made by the DM-master.
+// As long as the DM-worker can connect to the cluster, it must obey these decisions.
+// If the DM-worker can't connect to the cluster, it must shutdown all operations.
+//
+// In this model, we use etcd as the command queue for communication between the DM-master and DM-worker instead of gRPC.
+//
+// One example of the workflow:
+// 0. the user starts the DM-master cluster, and GET all history persisted data described above.
+//   - restore the in-memory status.
+// 1. the user starts a DM-worker instance.
+//   - PUT DM-worker instance information into etcd.
+// 2. DM-master GET the information of the DM-worker instance, and mark it as `free` status.
+// 3. the user adds an upstream config.
+//   - PUT the config of the upstream into etcd.
+// 4. DM-master schedules the upstream relevant operations to the free DM-worker.
+//   - PUT the bound relationship.
+//   - PUT the expectant stage of the relay if not exists.
+// 5. DM-worker GET the bound relationship, the config of the upstream and the expectant stage of the relay.
+// 6. DM-worker obey the expectant stage of the relay.
+//   - start relay (if error occurred, wait for the user to resolve it and do not re-schedule it to other DM-worker instances).
+// 7. the user starts a data migration task.
+// 8. DM-master PUT the data migration task config and the expectant stage of subtasks into etcd.
+// 9. DM-worker GET the config of the subtask, the expectant stage of the subtask.
+// 10. DM-worker obey the expectant stage of the subtask
+//   - start the subtask (if error occurred, wait for the user to resolve it).
+// 11. the task keeps running for a period.
+// 12. the user pauses the task.
+// 13. DM-master PUT the expectant stage of the subtask.
+// 14. DM-worker obey the expectant stage of the subtask.
+// 15. the user resumes the task (DM-master and DM-worker handle it similar to pause the task).
+// 16. the user stops the task.
+// 17. DM-master DELETE the data migration task config and the expectant stage of subtasks in etcd.
+//   - DELETE the information before subtasks shutdown.
+// 18. DM-worker stops the subtask.
+//   - NOTE: DM-worker should always stop the subtask if the expectant stage of the subtask is missing.
+// 19. the relay of the DM-worker continues to run.
+// 20. the user remove the upstream config.
+// 21. DM-master DELETE the upstream MySQL config, the bound relationship and the expectant stage of the relay.
+// 22. DM-worker shutdown.
+// 23. the user marks the DM-worker as offline.
+//   - DELETE DM-worker instance information in etcd.
+//
+// when the DM-worker (with relay and subtasks) is down:
+// 0. the status of the old DM-worker is un-health (keep-alive failed).
+// 1. DM-master choose another DM-worker instance for failover.
```

Comment:
maybe DM-master need to wait a while(two lease is enough? the old dm-worker need stop all subtask) and then schedule

Reply:
👍 you are right, I added a note in https://github.com/pingcap/dm/pull/473/commits/ca3191dd5c64461bce0c4c7c97de4869a151dc89.

We need to think about it before we know how long it is enough.